### PR TITLE
[GitHub] Advanced Sync rules

### PR DESCRIPTION
**Pull request by abhishekjoshi-crest**
## Closes https://github.com/elastic/connectors-py/issues/950

1. This PR introduces advanced filtering rules for GitHub by queries to filter out document based on that
2. Added documentation with several examples illustrating the syntax, how a basic use case works, limitations
3. Added documentation regarding impact of advanced sync rules on connector configuration and on document indexing
4. Included tooltip on the configurable fields that are ignored when advanced sync rules are used
5. As GitHub Documentation PR is not merged yet, we covered third point of the #950 in GitHub documentation PR.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [x] Considered corresponding documentation changes
- [x] Contributed any configuration settings changes to the configuration reference

### CODE REVIEW

File: docs/sync-rules/GITHUB.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+### Setting up the Github connector
+
+See the [Developer guide](../../docs/DEVELOPING.md) for setting up connectors.
+
+### Example advanced sync rules
+
+#### Schema for indexing document and files based on branch name configured via branch key
```
Comment (timgrein) [inline docs/sync-rules/GITHUB.md:7]:
```suggestion
#### Advanced rules for indexing document and files based on branch name configured via branch key
```

File: docs/sync-rules/GITHUB.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  {
+    "repository": "repo_name",
+    "filter": {
+      "issue": "is:bug"
+    }
+  }
+]
+```
+
+#### Schema for indexing document based on pr query related to open pr's via pr key
```
Comment (timgrein) [inline docs/sync-rules/GITHUB.md:33]:
```suggestion
#### Advanced rules for indexing document based on pr query related to open pr's via pr key 
```

File: connectors/sources/github.py [hunk 3]
```diff
@@ -1185,28 +1439,86 @@ async def get_docs(self, filtering=None):
...
+                        async for pull_request in self._fetch_pull_requests(
+                            repo_name=repo_name,
+                            response_key=("search",),
+                            filter_query=pull_request_query,
+                            query=GithubQuery.SEARCH_QUERY.value,
+                        ):
+                            yield pull_request, None
+                    else:
+                        self._logger.warning(
+                            f"Skipping pr for rule: {pull_request_query}"
```
[the remark this answers is not in the document]
Comment (abhishekjoshi-crest) [reply connectors/sources/github.py:1467]: `query_status` only returns boolean value whenever the query is not related to their object then it will return False otherwise True. Example: When issue query is passed in pr key then `query_status` will be False. IMO `query_status` shouldn't be part of logger as it will also False in skipping logger.

File: connectors/sources/github.py [hunk 4]
```diff
@@ -963,24 +1176,28 @@ async def _extract_pull_request(self, pull_request, owner, repo):
             pull_request.pop(field)
         yield pull_request
 
-    async def _fetch_pull_requests(self, repo_name):
+    async def _fetch_pull_requests(
+        self,
+        repo_name,
+        response_key=(REPOSITORY_OBJECT, "pullRequests"),
+        filter_query=None,
+        query=GithubQuery.PULL_REQUEST_QUERY.value,
```
[the remark this answers is not in the document]
Comment (abhishekjoshi-crest) [reply connectors/sources/github.py:1184]: Yup, will hide query via flag whether to fetch all object or fetch with custom query

File: connectors/sources/github.py [hunk 5]
```diff
@@ -373,6 +378,139 @@ class GithubQuery(Enum):
...
+                endCursor
+            }
+            nodes {
+                author {
+                login
+                }
+                body
+            }
+            }
+            reviews(first: 45) {
```
[the remark this answers is not in the document]
Comment (abhishekjoshi-crest) [reply connectors/sources/github.py:486]: In GitHub GraphQL maximum node limit is 500,000 and if we increase it this limit is exceeded.

File: connectors/sources/github.py [hunk 6]
```diff
@@ -231,47 +237,47 @@ class GithubQuery(Enum):
...
+        }}
+    }}
+    }}
     """
     COMMENT_QUERY = """
     query ($owner: String!, $name: String!, $number: Int!, $cursor: String) {{
     repository(owner: $owner, name: $name) {{
         {object_type}(number: $number) {{
-        comments(first: 100, after: $cursor) {{
+        comments(first: {node_size}, after: $cursor) {{
```
Comment (seanstory) [inline connectors/sources/github.py:280]: Hard to follow in this much diff, but should this be `NODE_SIZE` (capitalized)?

File: connectors/sources/github.py [hunk 7]
```diff
@@ -287,63 +293,63 @@ class GithubQuery(Enum):
...
+        }}
+    }}
+    }}
     """
     LABELS_QUERY = """
     query ($owner: String!, $name: String!, $number: Int!, $cursor: String) {{
     repository(owner: $owner, name: $name) {{
         {object_type}(number: $number) {{
-        labels(first: 100, after: $cursor) {{
+        labels(first: {node_size}, after: $cursor) {{
```
Comment (seanstory) [inline connectors/sources/github.py:352]: Another instance of missed caps?

File: connectors/sources/github.py [hunk 8]
```diff
@@ -360,7 +366,7 @@ class GithubQuery(Enum):
     query ($owner: String!, $name: String!, $number: Int!, $cursor: String) {{
     repository(owner: $owner, name: $name) {{
         {object_type}(number: $number) {{
-        assignees(first: 100, after: $cursor) {{
+        assignees(first: {node_size}, after: $cursor) {{
```
Comment (seanstory) [inline connectors/sources/github.py:369]: Another

File: connectors/sources/github.py [hunk 9]
```diff
@@ -533,22 +674,24 @@ async def get_github_item(self, resource):
...
-        """Get page information from the response.
+    def get_data_by_keys(self, response, keys, endKey):
+        """Retrieve data from a nested dictionary using a list of keys and an end key.
 
         Args:
-            response (dict): Response of GitHub
-            Keys (list): List of fields to get pageInfo
+            response (dict): The nested dictionary from which data will be extracted.
+            keys (list): A list of strings representing the keys to navigate the nested dictionary.
+            endKey (str): The final key to retrieve the desired data from the nested dictionary.
```
Comment (seanstory) [inline connectors/sources/github.py:683]: why have an `endKey` and not just have that be the last key in `keys`?

File: docs/sync-rules/GITHUB.md
Comment (seanstory) [inline docs/sync-rules/GITHUB.md:1]: This is really nice, having this guide alongside the code changes for review. Thanks for including these together.

File: docs/sync-rules/GITHUB.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    "filter": {
+      "issue": "is:bug",
+      "pr": "is:open",
+      "branch": "sync-rules-feature"
+    }
+  }
+]
+```
+
+**NOTE**: All documents pulled by a given rule are indexed regardless of whether the document has already been indexed by the previous rule. In such cases, document duplication won't happen but the count of indexed documents will be different in the logs. The correct count of documents can be referred from the ElasticSearch index.
```
Comment (seanstory) [inline docs/sync-rules/GITHUB.md:61]:
```suggestion
**NOTE**: All documents pulled by a given rule are indexed regardless of whether the document has already been indexed by the previous rule. In such cases, document duplication will happen, but the count of indexed documents will be different in the logs. The correct count of documents can be referred from the ElasticSearch index.
```

Because Elasticsearch will do the deduplication, right? So it does happen, just after sending duplicates to ES with the same IDs which overwrite one another.

File: tests/sources/test_github.py [hunk 11]
```diff
@@ -895,3 +902,233 @@ async def test_get_docs():
...
+                }
+            ],
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=ANY,
+            ),
+        ),
+        (
+            # invalid: invalid key
```
Comment (seanstory) [inline tests/sources/test_github.py:972]:
```suggestion
            # invalid: invalid array value
```

File: tests/sources/test_github.py [hunk 12]
```diff
@@ -895,3 +902,233 @@ async def test_get_docs():
...
+                }
+            ],
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=ANY,
+            ),
+        ),
+        (
+            # invalid: repository can be empty
```
Comment (seanstory) [inline tests/sources/test_github.py:988]:
```suggestion
            # invalid: repository can not be empty
```

File: tests/sources/test_github.py [hunk 13]
```diff
@@ -895,3 +902,233 @@ async def test_get_docs():
...
+            ),
+            [
+                REPOS,
+                MOCK_RESPONSE_PULL,
+                MOCK_RESPONSE_ISSUE,
+                MOCK_RESPONSE_ATTACHMENTS[0],
+            ],
+        ),
+        (
+            # Configured invalid queries, without branch
```
Comment (seanstory) [inline tests/sources/test_github.py:1079]:
```suggestion
            # Configured invalid queries, with branch
```

File: tests/sources/test_github.py
Comment (seanstory) [inline tests/sources/test_github.py:1]: great tests - those comments are super helpful

**Review by seanstory — COMMENTED** [review]
couple of nits, but largely looks good to me.

File: connectors/sources/github.py (hunk 6 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/github.py:280]: we will be using `object_type` and `node_size` in format string later so, I kept this variable name in smaller case and in other queries we are using f-string and directly assigning the `NODE_SIZE`. Same for `LABELS_QUERY` and `ASSIGNEES_QUERY`.

File: connectors/sources/github.py (hunk 9 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/github.py:683]: Because in `_fetch_pull_requests` and in `_fetch_issues` we are passing keys more than one time. one to get `pageInfo` and one to get `nodes`. So if we are remove the `endKey` then we need to update keys multiple times i.e. add `pageInfo` in keys then remove `pageInfo` and add `nodes` etc. So for simplicity I have added endKey argument.

File: connectors/sources/github.py (hunk 6 above)
Comment (seanstory) [reply connectors/sources/github.py:280]: Is it ever variable? Or are you always just passing that constant as the variable to the format?

**Review by seanstory — APPROVED** [review]

File: connectors/sources/github.py (hunk 6 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/github.py:280]: Yes we are passing constant in format because above query is not f-string as we also needs pass `object_type`.
